File: scaler_problems/day12-array-prefix-sum-intro/q3_closestMinMax.js
/**
 * Closest MinMax
Problem Description

Given an array A. Find the size of the smallest subarray such that it contains atleast one occurrence of the maximum value of the array
and atleast one occurrence of the minimum value of the array.

Problem Constraints
1 <= |A| <= 2000

Input Format
First and only argument is vector A

Output Format
Return the length of the smallest subarray which has atleast one occurrence of minimum and maximum element of the array

Example Input
Input 1:
A = [1, 3]
Input 2:
A = [2]

Example Output
Output 1:
 2
Output 2:
 1

Example Explanation
Explanation 1:
 Only choice is to take both elements.
Explanation 2:

 Take the whole array.
 */


//  TODO

const closestMinMax = (A) => {

    let min = 2000;
    let max = 1;
    for (let i = 0; i < A.length; i++) {
        if (A[i] > max) {
            max = A[i];
        }
        if (A[i] < min) {
            min = A[i];
        }
    }
    console.log('min', min);
    console.log('max', max);

    let minIndex = A.length - 1;
    let maxIndex = A.length - 1;
    let minSubArrayLength = A.length;

    for (let i = 0; i < A.length; i++) {
        if (A[i] === min) {
            minSubArrayLength = Math.min(minSubArrayLength, Math.abs(i - maxIndex) + 1);
            minIndex = i;
        }
        if (A[i] === max) {
            minSubArrayLength = Math.min(minSubArrayLength, Math.abs(i - minIndex) + 1);
            maxIndex = i;
        }
    }

    console.log('minSubArrayLength', minSubArrayLength)
}

closestMinMax([1, 3]);
closestMinMax([1, 3, 4, 5, 6, 3, 2, 1, 6, 2, 1]);
closestMinMax([377, 448, 173, 307, 108]);
closestMinMax([834, 563, 606, 221, 165]);

File: my_problems/day12-array-prefix-sum-intro/q3_Closest_min_max.java
/* Closest MinMax */

public class q3_Closest_min_max {

    public int solve(int[] A) {

        // initialize min and max index with -1
        int min_e = -1;
        int max_e = -1;
        // initialize answer with N as max length would be N
        int answer = A.length;

        // find min and max element from array
        int min = A[0];
        int max = A[0];
        for (int i = 1; i < A.length; i++) {
            min = Math.min(A[i], min);
            max = Math.max(A[i], max);
        }

        for (int i = 0; i < A.length; i++) {
            // if we find min, update min index
            if (A[i] == min) {
                min_e = i;
            }
            // if we find max, update max index
            if (A[i] == max) {
                max_e = i;
            }
            // update only if we have both the min and max indices > -1
            if (min_e > -1 && max_e > -1) {
                answer = Math.min(answer, Math.abs(min_e - max_e) + 1);
            }
        }
        return answer;

    }

    public static void main(String[] args) {

        q3_Closest_min_max t1 = new q3_Closest_min_max();
        int[] A;
        {
            A = new int[] { 1, 3 };
            System.out.println(t1.solve(A));
        }
        {
            A = new int[] { 1, 3, 4, 5, 6, 3, 2, 1, 6, 2, 1 };
            System.out.println(t1.solve(A));
        }
        {
            A = new int[] { 377, 448, 173, 307, 108 };
            System.out.println(t1.solve(A));
        }
        {
            A = new int[] { 834, 563, 606, 221, 165 };
            System.out.println(t1.solve(A));
        }

    }

}
